### PR TITLE
kvserver: fix reproposals test with pipelined writes

### LINKED ISSUE #106504: kvserver: add test to verify tryReproposeWithNewLeaseIndex behaviour
This PR https://github.com/cockroachdb/cockroach/pull/106502 removes a flaky integration test that verified reproposals with new lease index doesn't break pipelined writes (errors of this type are correctly handled below request evaluation).
Removed integration test was too high level to pinpoint this use case and precisely filter all other associated failure modes.
Test was originally introduced in https://github.com/cockroachdb/cockroach/pull/35261 and PR has context on what test verifies and motivation for it.

This reproposal functionality is not covered by any tests as of now.

Jira issue: CRDB-29593

### PR DESCRIPTION
**Pull request by pav-kv**
Before this commit, the test was a no-op reporting 0 metrics. This is due to `isOurCommand` function which assumed that the key increment request is always the first request in `BatchRequest`. With pipelined writes, this is not true.

A typical request is:
> QueryIntent ["00001-testing",/Min), Increment ["00001-testing",/Min)

In this commit, `isOutCommand` scans the `BatchRequest` to find the command.

Before:
```
observed 0 async write restarts, observed 0/0 injected aborts, 0 injected illegal lease applied indexes
commands reproposed (unchanged): 1
commands reproposed (new LAI): 0
```
After:
```
observed 69 async write restarts, observed 0/69 injected aborts, 366 injected illegal lease applied indexes
commands reproposed (unchanged): 1
commands reproposed (new LAI): 297
```

Fixes #106504
Touches #110551
Epic: none
Release note: none

### CODE REVIEW

File: pkg/kv/kvserver/replica_raft_test.go [hunk 1]
```diff
@@ -337,19 +343,29 @@ func testProposalsWithInjectedLeaseIndexAndReproposalError(t *testing.T, pipelin
...
 		// If we're not pipelined, we shouldn't be able to get an async write
 		// failure. This isn't testing anything about reproposals per se, rather
 		// it's validation that we're truly not doing pipelined writes.
 		require.Zero(t, observedAsyncWriteFailures.Load())
+		// All the injected reproposal errors should manifest to the transaction.
+		require.Equal(t, injectedReproposalErrors.Load(), observedReproposalErrors.Load())
 	}
+	// All incorrect lease indices should manifest either as a reproposal, or a
+	// failed reproposal (when an error is injected).
+	require.Equal(t, insertedIllegalLeaseIndex.Load(),
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_raft_test.go:367]: Some of these metric comparisons might (or not) be flaky if we throw in lease transfers and retries. I've stressed the test 10k times though, and did not find any flakes. So proposing to leave as is for now, and deal with it if it flakes.

File: pkg/kv/kvserver/replica_raft_test.go [hunk 2]
```diff
@@ -165,7 +165,12 @@ func testProposalsWithInjectedLeaseIndexAndReproposalError(t *testing.T, pipelin
 		if ba == nil {
 			return "", false // not local proposal
 		}
-		inc := ba.Requests[0].GetIncrement()
+		var inc *kvpb.IncrementRequest
+		for i := range ba.Requests {
+			if inc = ba.Requests[i].GetIncrement(); inc != nil {
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_raft_test.go:170]: nit: can use `inc, ok := BatchRequest.GetArg(kvpb.Increment)`.

File: pkg/kv/kvserver/replica_raft_test.go (hunk 1 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_raft_test.go:367]: Lease transfers shouldn't be an issue at least, `testContext` tests use a single replica afaik.

**Review by erikgrinaker — APPROVED** [review]
Nice find, thanks!

File: pkg/kv/kvserver/replica_raft_test.go (hunk 2 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_raft_test.go]: Done.

**Comment by pavelkalinnikov** [comment]
bors r=erikgrinaker

**Pull request merged**